custom-elements/custom-elements-registry/define.html

```diff
@@ -0,0 +1,230 @@
+    });
+  }, 'If one argument, throw a TypeError');
+  [
+    [ "undefined", undefined ],
+    [ "null", null ],
+    [ "object", {} ],
+    [ "string", "string" ],
+    [ "arrow function", () => {} ], // IsConstructor returns false for arrow functions
+    [ "method", ({ m() { } }).m ], // IsConstructor returns false for methods
+  ].forEach(function (t) {
+    test(function () {
+      assert_throws(new TypeError, function () {
+        window.customElements.define("test-define-ctor-" + t[0], t[1]);
+      });
+    }, 'If constructor is ' + t[0] + ', throw a TypeError');
+  });
+
+  // 2. If name is not a valid custom element name,
+  // then throw a SyntaxError and abort these steps.
+  var validCustomElementNames = [
+    // [a-z] (PCENChar)* '-' (PCENChar)*
+    // https://html.spec.whatwg.org/multipage/scripting.html#valid-custom-element-name
+    "a-",
+    "a-a",
+    "aa-",
+    "aa-a",
+    "a-.-_",
+    "a-0123456789",
+    "a-\u6F22\u5B57", // Two CJK Unified Ideographs
+    "a-\uD840\uDC0B", // Surrogate pair U+2000B
+  ];
+  var invalidCustomElementNames = [
+    undefined,
+    null,
+    "",
+    "a",
+    "A",
+    "A-",
+    "0-",
+    "a-A",
+    "a-Z",
+    "A-a",
+    // name must not be any of the hyphen-containing element names.
+    "annotation-xml",
+    "color-profile",
+    "font-face",
+    "font-face-src",
+    "font-face-uri",
+    "font-face-format",
+    "font-face-name",
+    "missing-glyph",
+  ];
+  validCustomElementNames.forEach(function (name) {
+    test(function () {
+      window.customElements.define(name, class {});
+    }, 'If name is ' + name + ', success');
+  });
+  invalidCustomElementNames.forEach(function (name) {
+    test(function () {
+      assert_throws("SyntaxError", function () {
+        window.customElements.define(name, class {});
+      });
+    }, 'If name is ' + name + ', throw a SyntaxError');
+  });
+
+  // 3. If this CustomElementsRegistry contains an entry with name name,
+  // then throw a NotSupportedError and abort these steps.
+  test(function () {
+    window.customElements.define("test-define-dup-name", class {});
+    assert_throws("NOT_SUPPORTED_ERR", function () {
+      window.customElements.define("test-define-dup-name", class {});
+    });
+  }, "If the name is already defined, throw a NotSupportedError");
+
+  // 4. If this CustomElementsRegistry contains an entry with constructor constructor,
+  // then throw a NotSupportedError and abort these steps.
+  test(function () {
+    class testDupCtor {};
+    window.customElements.define("test-define-dup-ctor", testDupCtor);
+    assert_throws("NOT_SUPPORTED_ERR", function () {
+      window.customElements.define("test-define-dup-ctor2", testDupCtor);
+    });
+  }, "If ctor is already defined, throw a NotSupportedError");
+
+  // 7.1. If extends is a valid custom element name,
+  // then throw a NotSupportedError.
+  validCustomElementNames.forEach(function (name) {
+    test(function () {
```

Comment:
I think it makes more sense to hoist the test out and have one test for name validity, which makes multiple assertions.